runtime/src/main/cpp/MemorySharedRefs.cpp

```diff
@@ -47,21 +83,20 @@ void KRefSharedHolder::init(ObjHeader* obj) {
   obj_ = obj;
 }
 
+template <ErrorHandlingPolicy errorHandlingPolicy>
 ObjHeader* KRefSharedHolder::ref() const {
-  if (auto* result = refOrNull())
-    return result;
-
-  throwIllegalSharingException(obj_);
-}
-
-ObjHeader* KRefSharedHolder::refOrNull() const {
-  if (!isRefAccessible()) {
+  if (!ensureRefAccessible<errorHandlingPolicy>(obj_, context_)) {
     return nullptr;
   }
+
   AdoptReferenceFromSharedVariable(obj_);
   return obj_;
 }
 
+template ObjHeader* KRefSharedHolder::ref<ErrorHandlingPolicy::kReturnDefault>() const;
+template ObjHeader* KRefSharedHolder::ref<ErrorHandlingPolicy::kThrowException>() const;
+template ObjHeader* KRefSharedHolder::ref<ErrorHandlingPolicy::kTerminate>() const;
```

Comment:
Is this explicit instantiation necessary?  See more below.

Reply:
Yep. Since the definition of `KRefSharedHolder::ref` is in this file, compilation of this file must produce all instantiations that might be used. Explicit instantiation makes sure this happens.

It is strange, however, that these instantiations are not generated by `KRefSharedHolder::refOr…` methods (possibly, because they are `inline`). But we're removing them anyway, so explicit instantiations are mandatory.